### PR TITLE
`CopySnippet`: convert to TypeScript (HDS-2689)

### PR DESCRIPTION
**Pull request by KristinLBradley**
### :pushpin: Summary

If merged, this PR converts the `CopySnippet` to TypeScript.

The CopySnippet depends upon:
* https://github.com/hashicorp/design-system/pull/2120

### :link: External links

* Jira ticket: [HDS-2689](https://hashicorp.atlassian.net/browse/HDS-2689)

***

### 👀 Component checklist

- ~~[ ] Percy was checked for any visual regression~~
- [x] A changelog entry was added via [Changesets](https://github.com/changesets/changesets) if needed (see [templates here](https://github.com/hashicorp/design-system/blob/main/wiki/Website-Changelog.md#templates-for-npm-packages))

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

[HDS-2689]: https://hashicorp.atlassian.net/browse/HDS-2689?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/components/src/components/hds/copy/snippet/index.ts [hunk 1]
```diff
@@ -15,9 +16,23 @@ export const SUCCESS_ICON = 'clipboard-checked';
 export const ERROR_ICON = 'clipboard-x';
 export const DEFAULT_STATUS = 'idle';
 
-export default class HdsCopySnippetIndexComponent extends Component {
+interface HdsCopySnippetSignature {
+  Args: {
```
Comment (didoo) [inline packages/components/src/components/hds/copy/snippet/index.ts:20]: `onError`, `onSuccess` and `text` should come from the `hds-clipboard` modifier (they're forwarded to it)

File: packages/components/src/components/hds/copy/snippet/index.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ import Component from '@glimmer/component';
 import { assert } from '@ember/debug';
 import { tracked } from '@glimmer/tracking';
 import { action } from '@ember/object';
+import type { HdsCopySnippetColors } from './types';
 
 export const DEFAULT_COLOR = 'primary';
 export const COLORS = ['primary', 'secondary'];
```
Comment (didoo) [inline packages/components/src/components/hds/copy/snippet/index.ts:13]: this should come from `HdsCopySnippetColorValues`

File: packages/components/src/components/hds/copy/snippet/index.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ import Component from '@glimmer/component';
 import { assert } from '@ember/debug';
 import { tracked } from '@glimmer/tracking';
 import { action } from '@ember/object';
+import type { HdsCopySnippetColors } from './types';
 
 export const DEFAULT_COLOR = 'primary';
```
Comment (didoo) [inline packages/components/src/components/hds/copy/snippet/index.ts:12]: this should come from `HdsCopySnippetColorValues.Primary`

**Review by didoo — COMMENTED** [review]
Left a few comments (similar to the `CopyButton`)

File: .changeset/popular-eels-melt.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"@hashicorp/design-system-components": minor
+---
+
+`CopyButton` - Converted component to TypeScript
```
Comment (KristinLBradley) [inline .changeset/popular-eels-melt.md:5]: I based this copySnippet branch off the copyButton branch originally. I've tried re-basing onto the newer clipboard branch but not sure how to fix these unrelated changes in here. (Maybe I need to keep it based off of the copyButton branch?)

Comment (didoo) [reply .changeset/popular-eels-melt.md:5]: I think you have different options
- just base off the `hds-2688-copy-button-ts-conversion` branch (the simplest)
- rename this branch `hds-2689-copy-snippet-ts-conversion__OLD`, create a new branch with the `hds-2689-copy-snippet-ts-conversion` name and cherry pick only the commits related to the `CopySnippet`, and then force push
- delete from the commit history of this branch the commits that are related to the `hds-clipboard` modifier or the `CopyButton` component, and then force push

let me know if you want we can do it together

Comment (KristinLBradley) [reply .changeset/popular-eels-melt.md]: I decided to keep it simple and just change it back to being based off of `hds-2688-copy-button-ts-conversion`. Thanks!

File: packages/components/src/components/hds/copy/snippet/index.ts [hunk 5]
```diff
@@ -7,17 +7,47 @@ import Component from '@glimmer/component';
...
 
-export default class HdsCopySnippetIndexComponent extends Component {
+interface HdsCopySnippetSignature {
+  Args: {
+    color?: HdsCopySnippetColors;
+    isFullWidth?: boolean;
+    isTruncated?: boolean;
+    onError?: HdsClipboardModifierSignature['Args']['Named']['onError'];
+    onSuccess?: HdsClipboardModifierSignature['Args']['Named']['onSuccess'];
+    textToCopy: HdsClipboardModifierSignature['Args']['Named']['text'];
```
Comment (didoo) [inline packages/components/src/components/hds/copy/snippet/index.ts:29]: see: https://helios.hashicorp.design/components/copy/snippet?tab=code#component-api
```suggestion
    color?: HdsCopySnippetColors;
    isFullWidth?: boolean;
    textToCopy: HdsClipboardModifierSignature['Args']['Named']['text'];
    isTruncated?: boolean;
    onSuccess?: HdsClipboardModifierSignature['Args']['Named']['onSuccess'];
    onError?: HdsClipboardModifierSignature['Args']['Named']['onError'];
```

File: packages/components/src/components/hds/copy/snippet/index.ts [hunk 6]
```diff
@@ -7,17 +7,47 @@ import Component from '@glimmer/component';
...
+export default class HdsCopySnippetComponent extends Component<HdsCopySnippetSignature> {
   @tracked status = DEFAULT_STATUS;
-  @tracked timer;
+  @tracked timer: ReturnType<typeof setTimeout> | undefined;
+
+  /**
+   * @param textToCopy
+   * @type {string | number | bigint | undefined} ???
+   */
+  get textToCopy() {
```
Comment (didoo) [inline packages/components/src/components/hds/copy/snippet/index.ts:42]: I want to discuss this with @alex-ju (I see you've done to overcome TS complaining about `bigint` not being assignable to type 'ContentValue'). I would be keen to drop support for `bigint` in the `hds-modifier` to avoid this extra layer of complexity (consumers can always easily convert a `bigint` to a `string`.

File: packages/components/src/components/hds/copy/snippet/index.ts [hunk 7]
```diff
@@ -102,23 +132,25 @@ export default class HdsCopySnippetIndexComponent extends Component {
   }
 
   @action
-  onSuccess(args) {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  onSuccess(args: any) {
```
Comment (didoo) [inline packages/components/src/components/hds/copy/snippet/index.ts:136]: same as for the `Copy::Button`
```suggestion
  onSuccess(args: HdsClipboardModifierSignature['Args']['Named']['onSuccess']) {
```

File: packages/components/src/components/hds/copy/snippet/index.ts [hunk 8]
```diff
@@ -102,23 +132,25 @@ export default class HdsCopySnippetIndexComponent extends Component {
...
 
     if (typeof onSuccess === 'function') {
       onSuccess(args);
     }
   }
 
   @action
-  onError(args) {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  onError(args: any) {
```
Comment (didoo) [inline packages/components/src/components/hds/copy/snippet/index.ts:149]: same as for the `Copy::Button`
```suggestion
  onError(args: HdsClipboardModifierSignature['Args']['Named']['onError']) {
```

File: packages/components/src/template-registry.ts [hunk 9]
```diff
@@ -157,6 +158,11 @@ export default interface HdsComponentsRegistry {
   'hds/copy/button': typeof HdsCopyButtonComponent;
   HdsCopyButton: typeof HdsCopyButtonComponent;
 
+  // Copy Snippet
+  'Hds::Copy::Snippet': typeof HdsCopySnippetComponent;
+  'hds/copy/snippet': typeof HdsCopySnippetComponent;
+  HdsCopySnippet: typeof HdsCopySnippetComponent;
```
Comment (didoo) [inline packages/components/src/template-registry.ts:164]: See: https://github.com/hashicorp/design-system/pull/2127
```suggestion
```

**Review by didoo — COMMENTED** [review]
The code is good and could be approved, but before I want to discuss with @alex-ju the issue with the `bigint` support.

File: packages/components/src/components/hds/copy/snippet/index.ts (hunk 6 above)
Comment (didoo) [reply packages/components/src/components/hds/copy/snippet/index.ts:42]: OK, I understand why the confusion (my confusion, I mean). the `@textToCopy` (argument) is passed down to the `hds-clipboard` modifier, while `this.textToCopy` is the one used to show the value at UI level. My suggestion is to rename the getter to avoid this confusion, something like this:
```suggestion
  get textToShow() {
```

**Review by didoo — APPROVED** [review]
Apart from the suggestions, everything else is OK 👍

**Comment by KristinLBradley** [comment]
I've been trying to rebase this branch onto [hds-2688-copy-button-ts-conversion](https://github.com/hashicorp/design-system/tree/hds-2688-copy-button-ts-conversion) but keep getting conflicts and when I try to fix the conflicts and continue rebasing the conflicts re-occur.

**Comment by didoo** [comment]
> @didoo I've been trying to rebase this branch onto [hds-2688-copy-button-ts-conversion](https://github.com/hashicorp/design-system/tree/hds-2688-copy-button-ts-conversion) but keep getting conflicts and when I try to fix the conflicts and continue rebasing the conflicts re-occur.

For future reference: this was fixed yesterday with a rebase of the branch `hds-2689-copy-snippet-ts-conversion` on `hds-2688-copy-button-ts-conversion` (and in the process of fixing conflicts I've merged and/or cleaned up a few commits)

**Review by alex-ju — APPROVED** [review]
